src/providers/aws.rs

```diff
@@ -1369,6 +1275,120 @@ impl RegionLauncher {
+                        .state
+                        .expect("spot request did not have state specified");
+                    let instance_id = sir.instance_id;
+                    (request_id, state, instance_id)
+                })
+                .collect();
+            break Ok(instances);
+        }
+    }
+
+    #[instrument(level = "debug")]
+    async fn cancel_spot_instance_requests(&self, request_ids: &Vec<String>) -> Result<(), Report> {
+        tracing::warn!(
+            "wait time exceeded for {} -- cancelling run",
+            self.region.name()
+        );
+        let mut cancel = rusoto_ec2::CancelSpotInstanceRequestsRequest::default();
+        cancel.spot_instance_request_ids = request_ids.clone();
+        self.client
+            .as_ref()
+            .unwrap()
+            .cancel_spot_instance_requests(cancel)
+            .await
+            .wrap_err("failed to cancel spot instances")?;
+
+        tracing::trace!("spot instances cancelled -- waiting for cancellation");
+        loop {
+            tracing::trace!("checking spot request status");
+            let instances = self.describe_spot_instance_requests(&request_ids).await?;
+
+            let all_cancelled = instances.iter().all(|(request_id, state, instance_id)| {
+                if state == "closed" || state == "cancelled" || state == "completed" {
+                    tracing::trace!(%request_id, %state, ?instance_id, "spot instance request cancelled");
+                    true
+                } else {
+                    false
+                }
+            });
+
+            if all_cancelled {
+                tracing::trace!("deleting spot instances");
+                // find instances with an id assigned and terminate them
+                let instance_ids = instances
+                    .into_iter()
+                    .filter_map(|(_, _, instance_id)| instance_id)
+                    .collect();
+                self.terminate_instances(instance_ids).await?;
+                break;
+            }
+
+            // let's not hammer the API
+            tokio::time::delay_for(time::Duration::from_secs(1)).await;
+        }
+        Ok(())
+    }
+
+    #[instrument(level = "debug")]
+    async fn terminate_instances(&self, instance_ids: Vec<String>) -> Result<(), Report> {
+        if instance_ids.is_empty() {
+            return Ok(());
+        }
+        let client = self.client.as_ref().unwrap();
+        let mut termination_req = rusoto_ec2::TerminateInstancesRequest::default();
+        termination_req.instance_ids = instance_ids;
+        while let Err(e) = client.terminate_instances(termination_req.clone()).await {
+            let msg = e.to_string();
+            if msg.contains("Pooled stream disconnected") || msg.contains("broken pipe") {
+                tracing::trace!("retrying instance termination");
+                continue;
+            } else {
+                // Why is `?` here ok? either:
```

Comment:
This comment should be on the `?` where `terminate_instances` is awaited, not here inside `terminate_instances`.